src/pages/Signup.tsx

```diff
@@ -0,0 +1,207 @@
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { sendSignUpData, fetchUserRoles, addData } from "../redux/Slice";
+import { useSelector, useDispatch } from "react-redux";
+import { BiLoader } from "react-icons/bi";
+import { faEye, faEyeSlash } from "@fortawesome/free-solid-svg-icons";
+import { AppDispatch } from "../redux/Store";
+
+const SignUp: React.FC = () => {
+  const [showPassword, setShowPassword] = useState<boolean>();
+  const [signUpFormData, setSignUpFormData] = useState({
+    firstName: "",
+    lastName: "",
+    email: "",
+    password: "",
+    roleId: "",
+  });
+  const [inputFieldError, setInputFieldError] = useState({
+    passwordError: true,
+    nameError: "",
+  });
+
+  const checkLengthOfPassword = /.{8,}/;
+  const upperCaseOfPassword = /[A-Z]/;
+  const lowerCaseOfPassword = /[a-z]/;
+  const digitOfPassword = /\d/;
+  const characterInPassword = /[!@#$%^&*]/;
+  const navigate = useNavigate();
+  const dispatch = useDispatch<AppDispatch>();
+  const userRole = useSelector((state: any) => state.data);
+  const signUpInfo = useSelector((state: any) => state);
+
+  const handleInput = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const updatedSignUpFormData = {
+      ...signUpFormData,
+      [event.target.id]: event.target.value,
+    };
+    let nameError = "";
+    if (updatedSignUpFormData.firstName.length < 5) {
+      nameError = "firstName";
+    } else if (updatedSignUpFormData.lastName.length < 5) {
+      nameError = "lastName";
+    }
+    setInputFieldError({
+      ...inputFieldError,
+      nameError: nameError,
+      passwordError: true,
+    });
+    setSignUpFormData(updatedSignUpFormData);
+  };
+  const handleSelect = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setSignUpFormData({
+      ...signUpFormData,
+      roleId: event.target.value,
+    });
+  };
+  useEffect(() => {
+    dispatch(fetchUserRoles());
+  }, []);
+
+  const handleSignUpSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    if (
+      !checkLengthOfPassword.test(signUpFormData.password) ||
+      !upperCaseOfPassword.test(signUpFormData.password) ||
+      !lowerCaseOfPassword.test(signUpFormData.password) ||
+      !digitOfPassword.test(signUpFormData.password) ||
+      !characterInPassword.test(signUpFormData.password)
+    ) {
+      setInputFieldError({
+        ...inputFieldError,
+        passwordError: false,
+      });
+      return;
+    }
+    if (inputFieldError.passwordError) {
+      const userData = {
+        firstName: signUpFormData.firstName,
+        lastName: signUpFormData.lastName,
+        email: signUpFormData.email,
+        password: signUpFormData.password,
+        roleId: signUpFormData.roleId,
+      };
+      dispatch(addData(userData));
+      dispatch(sendSignUpData(userData));
+    }
+    if (!signUpInfo.isError && !signUpInfo.isLoading) {
+      navigate("/");
```

Comment:
If Signup has any error why are you redirecting? you should be showing this error @Rishab-innovative